docs/source/custom-collections.rst

```diff
@@ -0,0 +1,326 @@
+Custom Collections
+==================
+
+For many problems the built-in dask collections (``dask.array``,
+``dask.dataframe``, ``dask.bag``, and ``dask.delayed``) are sufficient. For
+cases where they aren't it's possible to create your own dask collections. Here
+we describe the required methods to fullfill the dask collection interface.
+
+.. note:: This is considered an advanced feature. For most cases the built-in
+          collections are probably sufficient.
+
+Before reading this you should read and underestand:
+
+- :doc:`overview <graphs>`
+- :doc:`graph specification <spec>`
+- :doc:`custom graphs <custom-graphs>`
+
+The Dask Collection Interface
+-----------------------------
+
+To create your own dask collection, you need to fullfill the following
+interface. Note that there is no required base class:
+
+.. method:: __dask_graph__(self)
```

Comment:
Would it be better to name things like `_dask_graph_` instead? Double underscore invokes name mangling in some cases, and I'm not sure that's actually helpful here.

Also, although obviously Python is never going to name methods with the name "dask" in them, the double underscore prefix+postfix is technically reserved for language (I've gotten occasional push-back about this before, e.g., when I proposed an ad-hoc protocol for pandas on python-ideas).

Reply:
I'm guessing this is done to match what NumPy does with its array interface.

Double underscore before and after is different from just double underscore before (without the double underscore after). Naming mangling only applies to the latter case.

Reply:
Indeed, this was done to mirror how numpy and other libraries implement interfaces. I see no issue with using `__foo__` methods here, but happy to change if there's considerable pushback.

Personally I feel that private methods feel wrong here (usually indicate private interfaces that shouldn't be used), while dunder methods look more intentionally thought out.

Reply:
Am more than happy for us to mirror NumPy. :)